Refactor FXIOS-11799 [Profile] Decouple SearchEnginesManager from Profile

antigluten opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-11799)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/25731)

## :bulb: Description
Decoupled `SearchEnginesManager` from Profile

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Settings/CustomSearchViewController.swift:
```diff
@@ -68,7 +68,7 @@ class CustomSearchViewController: SettingsTableViewController {
                 let engine = try await createEngine(query: trimmedQuery, name: trimmedTitle)
                 self.spinnerView.stopAnimating()
-                self.profile?.searchEnginesManager.addSearchEngine(engine)
+                self.searchEnginesManager?.addSearchEngine(engine)
```
Nit: Could we please remove self, from all this Task, it should be not needed.

antigluten replied:
✅ Done

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Settings/Main/General/SearchSetting.swift:
```diff
@@ -28,7 +28,7 @@ class SearchSetting: Setting {
          settingsDelegate: GeneralSettingsDelegate?) {
-        self.profile = settings.profile
+        self.searchEnginesManager = settings.searchEnginesManager
         self.settingsDelegate = settingsDelegate
```
could we inject directly the search engine manager ? also could we remove the coupling with the `SettingsTableViewController` and inject directly the current theme ? what do you think ?

antigluten replied:
✅ Done

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Toolbar/AddressToolbarContainerModelTests.swift:
```diff
@@ -18,7 +18,11 @@ class AddressToolbarContainerModelTests: XCTestCase {
+        )
     }
 
     override func tearDown() {
```
nit: could we move the `super.tearDown()` at the end of the method. it is just a best practice

antigluten replied:
✅ Done

FilippoZazzeroni reviewed (changes requested):
Great work @antigluten just few small nits. thanks for your contribution 😃

antigluten commented on firefox-ios/Client/Frontend/Settings/SettingsTableViewController.swift:
```diff
@@ -824,6 +824,7 @@ class SettingsTableViewController: ThemedTableViewController {
 
     var profile: Profile?
     var tabManager: TabManager?
+    var searchEnginesManager: SearchEnginesManager?
```
✅ Done

FilippoZazzeroni reviewed (approved):
LGTM @antigluten 🎉 great job

antigluten commented:
@FilippoZazzeroni, I've fixed some test, could you review them, please? Also, is it okay to leave registering dependencies at that?

FilippoZazzeroni commented on firefox-ios/Client/Application/DependencyHelper.swift:
```diff
@@ -15,6 +15,12 @@ class DependencyHelper {
+        let searchEnginesManager = appDelegate.searchEnginesManager
+        AppContainer.shared.register(service: searchEnginesManager)
+
+        let searchEnginesManagerProvider: SearchEnginesManagerProvider = searchEnginesManager
```
We should register only one object, the other is a duplicate.

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Browser/SearchEngines/Redux/SearchEngineSelectionMiddleware.swift:
```diff
@@ -11,11 +11,11 @@ final class SearchEngineSelectionMiddleware {
 
     init(profile: Profile = AppContainer.shared.resolve(),
-         searchEnginesManager: SearchEnginesManagerProvider? = nil,
+         searchEnginesManager: SearchEnginesManagerProvider = AppContainer.shared.resolve(),
```
i think this is the correct implementation, unless we need a specific down cast to `SearchEngineProvider` we should everywhere register just the `SearchEngineManagerProvider` passing the `SearchEngineManager` as concrete object.

FilippoZazzeroni reviewed (changes requested):
I'd refactor the way we use `SearchEngineManagerProvider` and pass down it's concrete object masked with that protocol interface. Let me know what you think. thanks again for your contribution

antigluten commented:
> I'd refactor the way we use `SearchEngineManagerProvider` and pass down it's concrete object masked with that protocol interface. Let me know what you think. thanks again for your contribution

Refactored the way we use `SearchEnginesManager` -> `SearchEnginesManagerProvider` wherever possible

FilippoZazzeroni reviewed (changes requested):
I looked on the implementation, and i think we should rely directly on `SearchEngineManager` at this point. Sorry to see it right now, but this PR touches a lot of files. Anyway i'm telling this since we shouldn't have this case in the source code, as in the case the cast doesn't succeed then we are not presenting anything to the user.
```swift
func presentSearchSettingsController() {
        let searchSettingsTableViewController = SearchSettingsTableViewController(profile: profile, windowUUID: windowUUID)
        guard let searchEnginesManager = searchEnginesManager as? SearchEnginesManager else {  
// what happens here ?
return }
        let searchSettingsTableViewController = SearchSettingsTableViewController(
            profile: profile,
            searchEnginesManager: searchEnginesManager,
            windowUUID: windowUUID
        )
        let navController = ModalSettingsNavigationController(rootViewController: searchSettingsTableViewController)
        self.present(navController, animated: true, completion: nil)
    }
```

So as a general rule i think we should register in the `AppContainer` directly the `SearchEngineManager` instead of the protocol, and pass down the concrete type and where possible program at the interface so declare `SearchEngineManagerProvider` but pass down the concrete object.

if the interface needed by the class is not enough than we can declare it as the concrete object directly, and in future tickets we can improve the protocol to add/split the missing functionalities.
Last thing i see in some places we inject the manager via `AppContainer` and in some places not. Since we register this dependency trough the container, i'd inject it only through the AppContainer, so we are consistent with the rest of the implementation.

antigluten commented:
> I looked on the implementation, and i think we should rely directly on `SearchEngineManager` at this point. Sorry to see it right now, but this PR touches a lot of files. Anyway i'm telling this since we shouldn't have this case in the source code, as in the case the cast doesn't succeed then we are not presenting anything to the user.
> 
> ```swift
> func presentSearchSettingsController() {
>         let searchSettingsTableViewController = SearchSettingsTableViewController(profile: profile, windowUUID: windowUUID)
>         guard let searchEnginesManager = searchEnginesManager as? SearchEnginesManager else {  
> // what happens here ?
> return }
>         let searchSettingsTableViewController = SearchSettingsTableViewController(
>             profile: profile,
>             searchEnginesManager: searchEnginesManager,
>             windowUUID: windowUUID
>         )
>         let navController = ModalSettingsNavigationController(rootViewController: searchSettingsTableViewController)
>         self.present(navController, animated: true, completion: nil)
>     }
> ```
> 
> So as a general rule i think we should register in the `AppContainer` directly the `SearchEngineManager` instead of the protocol, and pass down the concrete type and where possible program at the interface so declare `SearchEngineManagerProvider` but pass down the concrete object.
> 
> if the interface needed by the class is not enough than we can declare it as the concrete object directly, and in future tickets we can improve the protocol to add/split the missing functionalities. Last thing i see in some places we inject the manager via `AppContainer` and in some places not. Since we register this dependency trough the container, i'd inject it only through the AppContainer, so we are consistent with the rest of the implementation.

✅ Done

FilippoZazzeroni commented on firefox-ios/Client/Frontend/Library/Bookmarks/Legacy/BookmarksFolderCell.swift:
```diff
@@ -91,13 +91,14 @@ extension BookmarkItemData: BookmarksFolderCell {
                    libraryPanelDelegate: LibraryPanelDelegate?,
                    navigationController: UINavigationController?,
                    logger: Logger) {
+        let searchEnginesManager: SearchEnginesManager = AppContainer.shared.resolve()
```
We should inject this, so we can enable testability with mock data.

antigluten replied:
✅ Done

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Frontend/Home/TopSites/TopSitesDataAdaptorTests.swift:
```diff
@@ -4,35 +4,41 @@
         profile = MockProfile(databasePrefix: "FxHomeTopSitesManagerTests")
         profile.reopen()
 
+        searchEnginesManager = AppContainer.shared.resolve()
```
i normally prefer to mock directly the object instead of relying of dependency injection. Here i would add 
```
searchEngineManager = MockSearchEngineManager(..) 
```

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/TelemetryWrapperTests.swift:
```diff
@@ -20,11 +23,14 @@ class TelemetryWrapperTests: XCTestCase {
         Glean.shared.resetGlean(clearStores: true)
         Experiments.events.clearEvents()
+
+        profile = AppContainer.shared.resolve()
```
same here, we should use `MockProfile` directly and if not needed we remove `DependencyHelperMock`

FilippoZazzeroni commented on firefox-ios/firefox-ios-tests/Tests/ClientTests/Toolbar/AddressToolbarContainerModelTests.swift:
```diff
@@ -16,35 +16,33 @@ class AddressToolbarContainerModelTests: XCTestCase {
-
-        // The MockProfile creates a SearchEnginesManager with a `MockSearchEngineProvider`
-        searchEnginesManager = mockProfile.searchEnginesManager
+        searchEnginesManager = AppContainer.shared.resolve()
```
same here `MockSearchEngineManager`

FilippoZazzeroni reviewed (changes requested):
Some nits about testing and mocks the rest LGTM great work @antigluten 🎉

FilippoZazzeroni reviewed (approved):
LGTM 🎉  great work @antigluten

antigluten commented:
@FilippoZazzeroni, I've fixed following tests:

- [x] `SearchEngineSelectionMiddlewareTests.testViewDidLoad_dispatchesDidLoadSearchEngines()`
- [ ] `TopSitesDataAdaptorTests.testSearchEngine_sponsoredTile_getsRemoved()`
- [x] `TopSitesViewModelTests.testDeletionOfAllDefaultSites()`
- [x] `TopSitesViewModelTests.testDeletionOfSingleSuggestedSite()`

Added a method to `AppContainer` that allows resolving dependencies by passing their concrete type. This is useful in scenarios like `SearchEngineSelectionMiddleware`, where we need to retrieve an instance from the DI container using its concrete type, while still maintaining the interface-based signature (`SearchEnginesManagerProvider`) in the middleware.

FilippoZazzeroni commented on BrowserKit/Sources/Common/DependencyInjection/AppContainer.swift:
```diff
@@ -16,8 +16,12 @@ public class AppContainer: ServiceProvider {
+        resolve(T.self)
+    }
+
+    public func resolve<T>(_ type: T.Type) -> T {
```
Looks good to me this, i would add if possible some documentation since this is our core business logics

antigluten replied:
I'd like to add some documentation by separate pull request

FilippoZazzeroni commented:
Thanks for fixing all the issue @antigluten we are almost there. I notice there is a merge conflict, we should resolve it prior running again the pipelines

antigluten commented:
@FilippoZazzeroni, looks like it's waiting for your re-approval

FilippoZazzeroni commented:
Hi @antigluten can you please rebase on main, i see Bitrise is failing, i cannot merge untill all the pipeline succeeds

antigluten commented:
> Hi @antigluten can you please rebase on main, i see Bitrise is failing, i cannot merge untill all the pipeline succeeds

Hi, done ✅
